src/main/java/com/google/sps/data/DatastoreAvailabilityDao.java

```diff
@@ -0,0 +1,176 @@
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Transaction;
+import com.google.appengine.api.datastore.TransactionOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.CompositeFilter;
+import com.google.appengine.api.datastore.Query.CompositeFilterOperator;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/** Accesses Datastore to support managing Availability entities. */
+public class DatastoreAvailabilityDao implements AvailabilityDao {
+  // @param datastore the DatastoreService we're using to interact with Datastore.
+  private DatastoreService datastore;
+
+  /** Initializes the fields for DatastoreAvailabilityDao. */
+  public DatastoreAvailabilityDao() {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  /**
+   * Retrieve the availability from Datastore from its id and wrap it in an Optional. If the id
+   * isn't in Datastore, the Optional is empty.
+   */
+  @Override
+  public Optional<Availability> get(long id) {
+    Key key = KeyFactory.createKey("Availability", id);
+    Entity availabilityEntity;
+    try {
+      availabilityEntity = datastore.get(key);
+    } catch (com.google.appengine.api.datastore.EntityNotFoundException e) {
+      return Optional.empty();
+    }
+    return Optional.of(entityToAvailability(availabilityEntity));
+  }
+
+  // Adds an Availability object into Datastore.
+  @Override
+  public void create(Availability avail) {
+    datastore.put(availabilityToNewEntity(avail));
+  }
+
+  // Updates the specified id with the new availability.
+  @Override
+  public void update(Availability avail) {
+    datastore.put(availabilityToUpdatedEntity(avail));
+  }
+
+  static Entity availabilityToNewEntity(Availability avail) {
+    Entity availabilityEntity = new Entity("Availability");
+    setEntityProperties(availabilityEntity, avail);
+    return availabilityEntity;
+  }
+
+  static Entity availabilityToUpdatedEntity(Availability avail) {
+    Entity availabilityEntity = new Entity("Availability", avail.id());
+    setEntityProperties(availabilityEntity, avail);
+    return availabilityEntity;
+  }
+
+  static void setEntityProperties(Entity availabilityEntity, Availability avail) {
+    availabilityEntity.setProperty("email", avail.email());
```

Comment:
After looking at another PR, l realize this function goes against something Alex warned me against. That is, a function that modifies a param instead of returning it. Here it might be fine because it's more explicit in the function name but not sure. @acmcarther  @jalexanderqed

Reply:
Or just combine availabilityToNewEntity and availabilityToUpdatedEntity.

Reply:
The Entity constructors used in these are different, so I guess I'll just put this code back within the two methods.